numpy/_core/tests/test_umath.py

```diff
@@ -2836,6 +2836,28 @@ def test_sign(self):
             assert_equal(res, tgt)
             assert_equal(out, tgt)
 
+    def test_sign_complex(self):
+        a = np.array([
+            np.inf, -np.inf, complex(0, np.inf), complex(0, -np.inf),
+            complex(np.inf, np.inf), complex(np.inf, -np.inf),  # nan
+            np.nan, complex(0, np.nan), complex(np.nan, np.nan),  # nan
+            0.0,  # 0.
+            3.0, -3.0, -2j, 3.0+4.0j, -8.0+6.0j
+        ])
+        out = np.zeros(a.shape, a.dtype)
+        tgt = np.array([
+            1., -1., 1j, -1j,
+            ] + [complex(np.nan, np.nan)] * 5 + [
```

Comment:
Is it perhaps confusing that this is the case:

```
In [1]: np.sign(np.nan)
Out[1]: np.float64(nan)

In [2]: np.sign(complex(np.nan))
Out[2]: np.complex128(nan+nanj)
```

Perhaps if the real or imaginary part is equal to zero, we should return 0 instead of NaN for the real or imaginary part of the result.

Reply:
Not sure, my sense would be that when either component is NaN, it is best to propagate that to both; it also keeps the result closest to the definition of `z/|z|` (for |z}->0,±inf, one can argue about meaningful limits, but less so for NaN).

Reply:
Thanks, on second thought I agree that having only a single kind of `nan` result here probably makes sense. I hope I never have a technical problem where complex NaN is something I need to worry about...